src/factor_graph.h

```diff
@@ -33,7 +33,10 @@ class FactorGraphDescriptor {
   /** number of edges */
   num_edges_t num_edges;
 
-  /** number of all weights */
+  /**
+   * number of weights. this is still kept here to maintain format of meta,
+   * and is also useful when loading weights
+   */
   num_weights_t num_weights;
```

Comment:
Here's another FactorGraph dependency with Weights. When calling InferenceResult's constructor, we pass in the FactorGraph and Weights array. InferenceResult preallocates the weights_{isfixed,value} arrays based on the FactorGraphDescriptor inside FactorGraph.

Again, this could be solved with creating a new LearningState class, I suppose? For now, I'll make things work with two floating information (num_weights and weights), and then later group num_weights and weights into a LearningState object. 

@netj I'm getting the sense that you don't like the std::vector idea?